File: swift/for_in.swift
class Section {
  let title: String
  let resetLessonPosition: Bool
  let lessons: [Lesson]
  var position: Int?

  init(_ title: String, _ resetLessonPosition: Bool, _ lessons: [Lesson]) {
    self.title = title
    self.resetLessonPosition = resetLessonPosition
    self.lessons = lessons
  }
}

class Lesson {
  let name: String
  var position: Int?
  init(_ name: String) {
    self.name = name
  }
}

var sections = [
  Section("Getting started", false,
          [Lesson("Welcome"),
           Lesson("Instalation")]),
  Section("Basic operator", false, [
          Lesson("Addition / Subtraction"),
          Lesson("Multiplication / Division")]),
  Section("Advanced topics", true, [
          Lesson("Mutability"),
          Lesson("Immutability")])
]

var lessonCounter = 1
var sectionCounter = 1

for section in sections {
  if section.resetLessonPosition {
    lessonCounter = 0
  }
  section.position = sectionCounter
  sectionCounter += 1
  for lesson in section.lessons {
    lesson.position = lessonCounter
    lessonCounter += 1
  }
}

debugPrint(sections)

File: nim/for_in.nim
import json

let sections = """[
  {
    "title": "Getting started",
    "reset_lesson_position": false,
    "lessons": [
      {"name": "Welcome"},
      {"name": "Installation"}
    ]
  },

  {
    "title": "Basic operator",
    "reset_lesson_position": false,
    "lessons": [
      {"name": "Addition / Subtraction"},
      {"name": "Multiplication / Division"}
    ]
  },

  {
    "title": "Advanced topics",
    "reset_lesson_position": true,
    "lessons": [
      {"name": "Mutability"},
      {"name": "Immutability"}
    ]
  }
]""".parseJson
var
  sectionCounter = 1
  lessonCounter = 1

for section in sections:
  if section["reset_lesson_position"].getBool:
    lessonCounter = 1

  section["position"] = sectionCounter.newJInt
  sectionCounter.inc()

  for lesson in section["lessons"].getElems:
    lesson["position"] = lessonCounter.newJInt
    lessonCounter.inc()

echo sections
